src/HttpOverStream.NamedPipe/NamedPipeListener.cs

```diff
@@ -31,26 +35,127 @@ public Task StartAsync(Action<Stream> onConnection, CancellationToken cancellati
+                    try
+                    {
+                        await serverStream.WaitForConnectionAsync(cancelToken).ConfigureAwait(false);
+
+                        // MP: We deliberately don't await this because we want to kick off the work on a background thead
+                        // and immediately check for the next client connecting
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+                        Task.Run(() => onConnection(serverStream));
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+
+                    }
+                    catch (OperationCanceledException) // Thrown when cancellationToken is cancelled
+                    {
+                        serverStream.Dispose();
+                    }
+                    catch (IOException ex) // Thrown if client disconnects early
+                    {
+                        if (ex.Message.Contains("The pipe is being closed."))
+                        {
+                            // TODO: Replace with logger
+                            Debug.WriteLine($"Could not read Named Pipe message - client disconnected before server finished reading.");
+                        }
+                        serverStream.Dispose();
+                    }
+                    catch (Exception)
+                    {
+                        // TODO: Maybe log
+                        serverStream.Dispose();
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                // TODO: Replace with logger
+                Debug.WriteLine("--- Exception creating server stream or waiting for connection: " + e);
+            }
+        }
+
+        private NamedPipeClientStream ConnectDummyClientAndServer(CancellationToken cancellationToken)
+        {
+            // Always have another stream active so if HandleStream finishes really quickly theres
+            // no chance of the named pipe being removed altogether.
+            // This is the same pattern as microsofts go library -> https://github.com/Microsoft/go-winio/pull/80/commits/ecd994be061f4ae21f463bbf08166d8edc96cadb
+            var serverStream = new NamedPipeServerStream(_pipeName, PipeDirection.InOut, _maxAllowedServerInstances, _pipeTransmissionMode, _pipeOptions);
+            serverStream.WaitForConnectionAsync(cancellationToken);
+            var dummyClientStream = new NamedPipeClientStream(".", _pipeName, PipeDirection.InOut, PipeOptions.Asynchronous);
+            dummyClientStream.Connect(100); // 100ms timeout to connect to itself
+
+            // We only need to return the client - Server will cancel + dispose itself via the token
```

Comment:
I wrote this comment yesterday, now I am doubting myself. Is this definitely the case?

Reply:
We need to keep track of both: if the server gets GC'd, it will close the native handle.